### PR TITLE
Websocket bugfixes

### LINKED ISSUE #1741: UnsafeRunSync in Websocket
So the unsafeRunSync in websockets can introduce a deadlock on machines with very few threads. This was observed with 2. To top if off the way we are handling that code with side-effects is just bad and I feel quite bad whenever I am reminded of it. 

We should look into refactoring this out into a better form where we don't do bad things.

[Section in Question](https://github.com/http4s/http4s/blob/master/blaze-core/src/main/scala/org/http4s/blazecore/websocket/Http4sWSStage.scala#L24-L25)

### LINKED ISSUE #2018: WSStage tests for blaze

### LINKED ISSUE #2022: websocket continues to send frames even after closed.

### PR DESCRIPTION
**Pull request by jmcardon**
Bugfixes all around

closes #2018 
closes #2022 
closes #1741 

Also: Squashed another bug where we did not send a close frame after receiving a close frame, which violated the websocket spec

### CODE REVIEW

File: .idea/codeStyles/Project.xml [hunk 1]
```diff
@@ -7,7 +7,6 @@
         </array>
       </option>
       <option name="sortAsScalastyle" value="true" />
-      <option name="USE_SCALAFMT_FORMATTER" value="true" />
```
Comment (jmcardon) [inline .idea/codeStyles/Project.xml]: Will fix when I update my idea version

File: blaze-core/src/main/scala/org/http4s/blazecore/websocket/Http4sWSStage.scala [hunk 2]
```diff
@@ -19,34 +19,83 @@ private[http4s] class Http4sWSStage[F[_]](ws: ws4s.Websocket[F])(
...
         channelRead().onComplete {
           case Success(ws) =>
             ws match {
-              case c @ Close(_) =>
-                unsafeRunSync[F, Unit](deadSignal.set(true))
-                cb(Right(c)) // With Dead Signal Set, Return callback with the Close Frame
+              case c: Close =>
+                if (sentClose.get()) {
+                  //We sent the close signal, thus we can be sure to end the stream here.
+                  F.toIO(deadSignal.set(true)).unsafeRunSync()
```
Comment (jmcardon) [inline blaze-core/src/main/scala/org/http4s/blazecore/websocket/Http4sWSStage.scala]: These were done as unsafeRunSync calls, due to issues with tests.

Originally, they were

```scala
F.toIO(deadSignal.set(true)).unsafeRunAsync(_ => cb(Right(c))
```

File: blaze-core/src/test/scala/org/http4s/blazecore/websocket/Http4sWSStageSpec.scala [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+
+    "send a close frame back and call the on close handler upon receiving a close frame" in {
+      val socket = TestSocket()
+      socket.sendInbound(Close())
+      socket.pollOutbound() must beSome(Close())
+      socket.pollOutbound(1) must beNone
+      socket.wasCloseHookCalled() must_== true
+    }
+
+    "not send two close frames " in {
```
Comment (jmcardon) [inline blaze-core/src/test/scala/org/http4s/blazecore/websocket/Http4sWSStageSpec.scala]: Test fails intermittently?

**Comment by ChristopherDavenport** [comment]
```
[error] /home/travis/build/http4s/http4s/blaze-core/src/test/scala/org/http4s/blazecore/websocket/WSTestHead.scala:47:12: value unit is not a member of object scala.concurrent.Future
[error]     Future.unit
[error]            ^
[error] one error found
```

Quick fix for 2.11

File: .idea/codeStyles/Project.xml [hunk 4]
```diff
@@ -1,16 +1,16 @@
...
+                    <option value="all other imports" />
+                </array>
+            </option>
+            <option name="sortAsScalastyle" value="true" />
+            <option name="USE_SCALAFMT_FORMATTER" value="true" />
+        </ScalaCodeStyleSettings>
+        <XML>
+            <option name="XML_LEGACY_SETTINGS_IMPORTED" value="true" />
+        </XML>
+    </code_scheme>
```
[the remark this answers is not in the document]
Comment (jmcardon) [reply .idea/codeStyles/Project.xml]: I did... no dice.

File: blaze-server/src/main/scala/org/http4s/server/blaze/WebSocketSupport.scala [hunk 5]
```diff
@@ -63,9 +65,12 @@ private[blaze] trait WebSocketSupport[F[_]] extends Http1ServerStage[F] {
                 case Success(_) =>
                   logger.debug("Switching pipeline segments for websocket")
 
-                  val segment = LeafBuilder(new Http4sWSStage[F](wsContext.webSocket))
-                    .prepend(new WSFrameAggregator)
-                    .prepend(new WebSocketDecoder)
+                  val deadSignal = F.toIO(Signal[F, Boolean](false)).unsafeRunSync()
```
[the remark this answers is not in the document]
Comment (jmcardon) [reply blaze-server/src/main/scala/org/http4s/server/blaze/WebSocketSupport.scala]: I was getting concurrent initialization issues for testing, which gave me inconsistent results (as a fun experiment, move that deadSignal to the field of the class, and see how many tests break. This gave me hours of frustration.)

if `stageShutDown()` didn't need it, we could avoid this entire call altogether.

File: blaze-core/src/main/scala/org/http4s/blazecore/websocket/Http4sWSStage.scala [hunk 6]
```diff
@@ -92,20 +141,12 @@ private[http4s] class Http4sWSStage[F[_]](ws: ws4s.Websocket[F])(
   }
 
   override protected def stageShutdown(): Unit = {
-    unsafeRunSync[F, Unit](deadSignal.set(true))
+    F.toIO(deadSignal.set(true)).unsafeRunSync()
```
[the remark this answers is not in the document]
Comment (jmcardon) [reply blaze-core/src/main/scala/org/http4s/blazecore/websocket/Http4sWSStage.scala]: I've done so much griefing over how to fix this but there's just no other way to shut the websocket down while running, from what I understand.

File: blaze-server/src/main/scala/org/http4s/server/blaze/WebSocketSupport.scala (hunk 5 above)
[the remark this answers is not in the document]
Comment (rossabaker) [reply blaze-server/src/main/scala/org/http4s/server/blaze/WebSocketSupport.scala]: That works for me.

**Comment by jmcardon** [comment]
CI Error over unrelated timeouts :(.

**Comment by ChristopherDavenport** [comment]
The fact this is reliably timeing out is worrisone.

**Comment by jmcardon** [comment]
```scala
[tut] compiling: /home/jose/work/http4s/docs/src/main/tut/hsts.md
[error] scala.MatchError: Some(-ross) (of class scala.Some)
[error]         at org.http4s.build.Http4sPlugin$.milestone$1(Http4sPlugin.scala:251)
[error]         at org.http4s.build.Http4sPlugin$.$anonfun$latestPerMinorVersion$4(Http4sPlugin.scala:257)
[error]         at scala.math.Ordering$$anon$5.compare(Ordering.scala:122)
[error]         at scala.math.Ordering.gteq(Ordering.scala:91)
[error]         at scala.math.Ordering.gteq$(Ordering.scala:91)
[error]         at scala.math.Ordering$$anon$5.gteq(Ordering.scala:121)
[error]         at scala.math.Ordering.max(Ordering.scala:103)
[error]         at scala.math.Ordering.max$(Ordering.scala:103)
[error]         at scala.math.Ordering$$anon$5.max(Ordering.scala:121)
[error]         at org.http4s.build.Http4sPlugin$.$anonfun$latestPerMinorVersion$6(Http4sPlugin.scala:259)
[error]         at scala.Option.fold(Option.scala:158)
[error]         at org.http4s.build.Http4sPlugin$.$anonfun$latestPerMinorVersion$1(Http4sPlugin.scala:259)
[error]         at scala.collection.IndexedSeqOptimized.foldLeft(IndexedSeqOptimized.scala:56)
[error]         at scala.collection.IndexedSeqOptimized.foldLeft$(IndexedSeqOptimized.scala:64)
[error]         at scala.collection.mutable.ArrayBuffer.foldLeft(ArrayBuffer.scala:48)
[error]         at org.http4s.build.Http4sPlugin$.latestPerMinorVersion(Http4sPlugin.scala:246)
[error]         at org.http4s.build.Http4sPlugin$.$anonfun$projectSettings$14(Http4sPlugin.scala:117)
[error]         at org.http4s.build.Http4sPlugin$.$anonfun$projectSettings$14$adapted(Http4sPlugin.scala:113)
[error]         at scala.Function1.$anonfun$compose$1(Function1.scala:44)
[error]         at sbt.internal.util.$tilde$greater.$anonfun$$u2219$1(TypeFunctions.scala:40)
[error]         at sbt.std.Transform$$anon$4.work(System.scala:67)
[error]         at sbt.Execute.$anonfun$submit$2(Execute.scala:269)
[error]         at sbt.internal.util.ErrorHandling$.wideConvert(ErrorHandling.scala:16)
[error]         at sbt.Execute.work(Execute.scala:278)
[error]         at sbt.Execute.$anonfun$submit$1(Execute.scala:269)
[error]         at sbt.ConcurrentRestrictions$$anon$4.$anonfun$submitValid$1(ConcurrentRestrictions.scala:178)
[error]         at sbt.CompletionService$$anon$2.call(CompletionService.scala:37)
[error]         at java.util.concurrent.FutureTask.run(FutureTask.java:266)
[error]         at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
[error]         at java.util.concurrent.FutureTask.run(FutureTask.java:266)
[error]         at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
[error]         at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
[error]         at java.lang.Thread.run(Thread.java:748)
[error] (docs / http4sBuildData) scala.MatchError: Some(-ross) (of class scala.Some)
```

File: blaze-core/src/main/scala/org/http4s/blazecore/websocket/Http4sWSStage.scala [hunk 7]
```diff
@@ -15,56 +15,105 @@ import org.http4s.websocket.WebsocketBits._
...
+    *
+    * - If we receive a ping frame, we MUST reply with a pong frame
+    * - If we receive a pong frame, we don't need to forward it.
+    * - If we receive a close frame, it means either one of two things:
+    *   - We sent a close frame prior, meaning we do not need to reply with one. Just end the stream
+    *   - We are the first to receive a close frame, so we try to atomically check a boolean flag,
+    *     to prevent sending two close frames. Regardless, we set the signal for termination of
+    *     the stream afterwards
+    *
+    * @return A websocket frame. Or an error. Only god knows
```
[the remark this answers is not in the document]
Comment (rossabaker) [reply blaze-core/src/main/scala/org/http4s/blazecore/websocket/Http4sWSStage.scala]: Looks like this is the last unresolved one, and I think we're good to go.

**Pull request merged**